mojap_metadata/converters/glue_converter/__init__.py

```diff
@@ -462,6 +529,37 @@ def generate_from_meta(
         boto_dict = self.gc.generate_from_meta(
             metadata, database_name=database_name, table_location=table_location
         )
+
+        # adding primary key to table parameters
+        if metadata.primary_key:
+            primary_key_dict = {"primary_key": str(metadata.primary_key)}
+            boto_dict["TableInput"]["Parameters"].update(primary_key_dict)
+
+        # protected attributes that should not be overwritten
+        additional_glue_table_properties = ["primary_key"]
+
+        table_properties_protected = (
+            _glue_table_properties_aws + additional_glue_table_properties
+        )
+
+        # updating glue table properties
+        metadata_dict = metadata.to_dict()
+
+        # nothing to do if glue_table_properties_custom are not in the schema
+        if "glue_table_properties_custom" not in metadata_dict.keys():
+            pass
+        # remove any protected glue table properties from the
+        # glue_table_properties_custom defined by the user in the metadata
+        else:
+            glue_table_properties_custom = {
+                k: str(v)
```

Comment:
I was just noseying through dbt-athena and they have [this function](https://github.com/dbt-athena/dbt-athena/blob/main/dbt/adapters/athena/utils.py#L16-L27)  (used [here](https://github.com/dbt-athena/dbt-athena/blob/main/dbt/adapters/athena/impl.py#L986C1-L987C1)) to convert the properties to strings.

Its got me wondering if we should do something similar here and store all the properties as valid json strings. Then when reading them back we could use `json.loads()` rather than `ast.literal_eval()`.

Appreciate we've been back and forth on this a bit so happy to stick with the current approach if we don't think that change is worth it, or theres something i'm missing.

Reply:
@davidbridgwood nice find, I do like the look of this!

I tried implementing it and it seems to work for `generate_from_meta`. 

However for `generate_to_meta`,  json.loads() errors when the type is a string e.g. `"csv"`, `"True"` as a valid json string should be enclosed in quotes e.g. `'"csv"'`, `'"True"'`. This won't be an issue if the table properties have been written using the updated `generate_from_meta` as I've updated the dbt-athena function to handle booleans too. However for existing table properties which have boolean values, these will be returned as a string. So it depends if we are okay with this? 

With `ast.literal_eval()` it will be able to handle existing table properties. 

Additionally, the table properties in the Glue Catalog will look like this if we use json.
![Screenshot 2024-05-15 at 13 18 26](https://github.com/moj-analytical-services/mojap-metadata/assets/123457715/74ee8e25-19cd-4ef8-a921-4a079d31e4d3)

Happy to have a discussion on what others think.

Reply:
Since we're moving to using strings in glue table properties, this is not a viable solution.